File: archive/pulumi-ts/Pr7149/lib/network-stack.ts
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';

export interface NetworkStackArgs {
  environmentSuffix: string;
  tags: pulumi.Input<{ [key: string]: string }>;
}

export class NetworkStack extends pulumi.ComponentResource {
  public readonly vpcId: pulumi.Output<string>;
  public readonly publicSubnetIds: pulumi.Output<string>[];
  public readonly privateSubnetIds: pulumi.Output<string>[];

  constructor(
    name: string,
    args: NetworkStackArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('tap:network:NetworkStack', name, args, opts);

    const { environmentSuffix, tags } = args;

    // Create VPC
    const vpc = new aws.ec2.Vpc(
      `payment-vpc-${environmentSuffix}`,
      {
        cidrBlock: '10.0.0.0/16',
        enableDnsHostnames: true,
        enableDnsSupport: true,
        tags: pulumi.output(tags).apply(t => ({
          ...t,
          Name: `payment-vpc-${environmentSuffix}`,
        })),
      },
      { parent: this }
    );

    // Get availability zones
    const azs = aws.getAvailabilityZones({ state: 'available' });

    // Create Internet Gateway
    const igw = new aws.ec2.InternetGateway(
      `payment-igw-${environmentSuffix}`,
      {
        vpcId: vpc.id,
        tags: pulumi.output(tags).apply(t => ({
          ...t,
          Name: `payment-igw-${environmentSuffix}`,
        })),
      },
      { parent: this }
    );

    // Create public subnets (3 AZs)
    const publicSubnets: aws.ec2.Subnet[] = [];
    for (let i = 0; i < 3; i++) {
      const subnet = new aws.ec2.Subnet(
        `payment-public-subnet-${i + 1}-${environmentSuffix}`,
        {
          vpcId: vpc.id,
          cidrBlock: `10.0.${i}.0/24`,
          availabilityZone: azs.then(az => az.names[i]),
          mapPublicIpOnLaunch: true,
          tags: pulumi.output(tags).apply(t => ({
            ...t,
            Name: `payment-public-subnet-${i + 1}-${environmentSuffix}`,
            Tier: 'Public',
          })),
        },
        { parent: this }
      );
      publicSubnets.push(subnet);
    }

    // Create private subnets (3 AZs)
    const privateSubnets: aws.ec2.Subnet[] = [];
    for (let i = 0; i < 3; i++) {
      const subnet = new aws.ec2.Subnet(
        `payment-private-subnet-${i + 1}-${environmentSuffix}`,
        {
          vpcId: vpc.id,
          cidrBlock: `10.0.${i + 10}.0/24`,
          availabilityZone: azs.then(az => az.names[i]),
          tags: pulumi.output(tags).apply(t => ({
            ...t,
            Name: `payment-private-subnet-${i + 1}-${environmentSuffix}`,
            Tier: 'Private',
          })),
        },
        { parent: this }
      );
      privateSubnets.push(subnet);
    }

    // Create Elastic IPs for NAT Gateways
    const eips: aws.ec2.Eip[] = [];
    for (let i = 0; i < 3; i++) {
      const eip = new aws.ec2.Eip(
        `payment-nat-eip-${i + 1}-${environmentSuffix}`,
        {
          domain: 'vpc',
          tags: pulumi.output(tags).apply(t => ({
            ...t,
            Name: `payment-nat-eip-${i + 1}-${environmentSuffix}`,
          })),
        },
        { parent: this }
      );
      eips.push(eip);
    }

    // Create NAT Gateways (one per AZ)
    const natGateways: aws.ec2.NatGateway[] = [];
    for (let i = 0; i < 3; i++) {
      const nat = new aws.ec2.NatGateway(
        `payment-nat-${i + 1}-${environmentSuffix}`,
        {
          subnetId: publicSubnets[i].id,
          allocationId: eips[i].id,
          tags: pulumi.output(tags).apply(t => ({
            ...t,
            Name: `payment-nat-${i + 1}-${environmentSuffix}`,
          })),
        },
        { parent: this }
      );
      natGateways.push(nat);
    }

    // Create public route table
    const publicRouteTable = new aws.ec2.RouteTable(
      `payment-public-rt-${environmentSuffix}`,
      {
        vpcId: vpc.id,
        tags: pulumi.output(tags).apply(t => ({
          ...t,
          Name: `payment-public-rt-${environmentSuffix}`,
        })),
      },
      { parent: this }
    );

    // Route to Internet Gateway
    new aws.ec2.Route(
      `payment-public-route-${environmentSuffix}`,
      {
        routeTableId: publicRouteTable.id,
        destinationCidrBlock: '0.0.0.0/0',
        gatewayId: igw.id,
      },
      { parent: this }
    );

    // Associate public subnets with public route table
    publicSubnets.forEach((subnet, i) => {
      new aws.ec2.RouteTableAssociation(
        `payment-public-rta-${i + 1}-${environmentSuffix}`,
        {
          subnetId: subnet.id,
          routeTableId: publicRouteTable.id,
        },
        { parent: this }
      );
    });

    // Create private route tables (one per AZ for NAT Gateway)
    privateSubnets.forEach((subnet, i) => {
      const privateRouteTable = new aws.ec2.RouteTable(
        `payment-private-rt-${i + 1}-${environmentSuffix}`,
        {
          vpcId: vpc.id,
          tags: pulumi.output(tags).apply(t => ({
            ...t,
            Name: `payment-private-rt-${i + 1}-${environmentSuffix}`,
          })),
        },
        { parent: this }
      );

      new aws.ec2.Route(
        `payment-private-route-${i + 1}-${environmentSuffix}`,
        {
          routeTableId: privateRouteTable.id,
          destinationCidrBlock: '0.0.0.0/0',
          natGatewayId: natGateways[i].id,
        },
        { parent: this }
      );

      new aws.ec2.RouteTableAssociation(
        `payment-private-rta-${i + 1}-${environmentSuffix}`,
        {
          subnetId: subnet.id,
          routeTableId: privateRouteTable.id,
        },
        { parent: this }
      );
    });

    // Create VPC Endpoints for S3
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const s3Endpoint = new aws.ec2.VpcEndpoint(
      `payment-s3-endpoint-${environmentSuffix}`,
      {
        vpcId: vpc.id,
        serviceName: 'com.amazonaws.us-east-1.s3',
        vpcEndpointType: 'Gateway',
        tags: pulumi.output(tags).apply(t => ({
          ...t,
          Name: `payment-s3-endpoint-${environmentSuffix}`,
        })),
      },
      { parent: this }
    );

    // Security group for VPC endpoints
    const endpointSecurityGroup = new aws.ec2.SecurityGroup(
      `payment-endpoint-sg-${environmentSuffix}`,
      {
        vpcId: vpc.id,
        description: 'Security group for VPC endpoints',
        ingress: [
          {
            protocol: 'tcp',
            fromPort: 443,
            toPort: 443,
            cidrBlocks: [vpc.cidrBlock],
            description: 'HTTPS from VPC',
          },
        ],
        egress: [
          {
            protocol: '-1',
            fromPort: 0,
            toPort: 0,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow all outbound',
          },
        ],
        tags: pulumi.output(tags).apply(t => ({
          ...t,
          Name: `payment-endpoint-sg-${environmentSuffix}`,
        })),
      },
      { parent: this }
    );

    // VPC Endpoint for ECR API
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const ecrApiEndpoint = new aws.ec2.VpcEndpoint(
      `payment-ecr-api-endpoint-${environmentSuffix}`,
      {
        vpcId: vpc.id,
        serviceName: 'com.amazonaws.us-east-1.ecr.api',
        vpcEndpointType: 'Interface',
        privateDnsEnabled: true,
        subnetIds: privateSubnets.map(s => s.id),
        securityGroupIds: [endpointSecurityGroup.id],
        tags: pulumi.output(tags).apply(t => ({
          ...t,
          Name: `payment-ecr-api-endpoint-${environmentSuffix}`,
        })),
      },
      { parent: this }
    );

    // VPC Endpoint for ECR Docker
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const ecrDkrEndpoint = new aws.ec2.VpcEndpoint(
      `payment-ecr-dkr-endpoint-${environmentSuffix}`,
      {
        vpcId: vpc.id,
        serviceName: 'com.amazonaws.us-east-1.ecr.dkr',
        vpcEndpointType: 'Interface',
        privateDnsEnabled: true,
        subnetIds: privateSubnets.map(s => s.id),
        securityGroupIds: [endpointSecurityGroup.id],
        tags: pulumi.output(tags).apply(t => ({
          ...t,
          Name: `payment-ecr-dkr-endpoint-${environmentSuffix}`,
        })),
      },
      { parent: this }
    );

    // VPC Endpoint for CloudWatch Logs
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const logsEndpoint = new aws.ec2.VpcEndpoint(
      `payment-logs-endpoint-${environmentSuffix}`,
      {
        vpcId: vpc.id,
        serviceName: 'com.amazonaws.us-east-1.logs',
        vpcEndpointType: 'Interface',
        privateDnsEnabled: true,
        subnetIds: privateSubnets.map(s => s.id),
        securityGroupIds: [endpointSecurityGroup.id],
        tags: pulumi.output(tags).apply(t => ({
          ...t,
          Name: `payment-logs-endpoint-${environmentSuffix}`,
        })),
      },
      { parent: this }
    );

    this.vpcId = vpc.id;
    this.publicSubnetIds = publicSubnets.map(s => s.id);
    this.privateSubnetIds = privateSubnets.map(s => s.id);

    this.registerOutputs({
      vpcId: this.vpcId,
      publicSubnetIds: this.publicSubnetIds,
      privateSubnetIds: this.privateSubnetIds,
    });
  }
}

File: archive/cdktf-py/Pr7948/lib/stacks/network_stack.py
"""Network infrastructure stack - VPC, subnets, security groups"""

from constructs import Construct
from cdktf_cdktf_provider_aws.vpc import Vpc
from cdktf_cdktf_provider_aws.subnet import Subnet
from cdktf_cdktf_provider_aws.internet_gateway import InternetGateway
from cdktf_cdktf_provider_aws.eip import Eip
from cdktf_cdktf_provider_aws.nat_gateway import NatGateway
from cdktf_cdktf_provider_aws.route_table import RouteTable, RouteTableRoute
from cdktf_cdktf_provider_aws.route_table_association import RouteTableAssociation
from cdktf_cdktf_provider_aws.security_group import SecurityGroup, SecurityGroupIngress, SecurityGroupEgress


class NetworkStack(Construct):
    """Creates VPC with 3 AZs, public and private subnets"""

    def __init__(self, scope: Construct, construct_id: str, region: str, environment_suffix: str):
        super().__init__(scope, construct_id)

        self.region = region
        self.environment_suffix = environment_suffix

        # VPC
        self.vpc = Vpc(
            self,
            "vpc",
            cidr_block="10.0.0.0/16",
            enable_dns_hostnames=True,
            enable_dns_support=True,
            tags={
                "Name": f"dr-vpc-{region}-{environment_suffix}"
            }
        )

        # Internet Gateway
        self.igw = InternetGateway(
            self,
            "igw",
            vpc_id=self.vpc.id,
            tags={
                "Name": f"dr-igw-{region}-{environment_suffix}"
            }
        )

        # Get availability zones (3 AZs)
        azs = [f"{region}a", f"{region}b", f"{region}c"]

        # Public subnets
        self.public_subnets = []
        for i, az in enumerate(azs):
            subnet = Subnet(
                self,
                f"public-subnet-{i}",
                vpc_id=self.vpc.id,
                cidr_block=f"10.0.{i}.0/24",
                availability_zone=az,
                map_public_ip_on_launch=True,
                tags={
                    "Name": f"dr-public-subnet-{i}-{region}-{environment_suffix}",
                    "Type": "public"
                }
            )
            self.public_subnets.append(subnet)

        # Private subnets
        self.private_subnets = []
        for i, az in enumerate(azs):
            subnet = Subnet(
                self,
                f"private-subnet-{i}",
                vpc_id=self.vpc.id,
                cidr_block=f"10.0.{i+10}.0/24",
                availability_zone=az,
                map_public_ip_on_launch=False,
                tags={
                    "Name": f"dr-private-subnet-{i}-{region}-{environment_suffix}",
                    "Type": "private"
                }
            )
            self.private_subnets.append(subnet)

        # Public route table
        self.public_route_table = RouteTable(
            self,
            "public-route-table",
            vpc_id=self.vpc.id,
            route=[RouteTableRoute(
                cidr_block="0.0.0.0/0",
                gateway_id=self.igw.id
            )],
            tags={
                "Name": f"dr-public-rt-{region}-{environment_suffix}"
            }
        )

        # Associate public subnets with public route table
        for i, subnet in enumerate(self.public_subnets):
            RouteTableAssociation(
                self,
                f"public-rt-assoc-{i}",
                subnet_id=subnet.id,
                route_table_id=self.public_route_table.id
            )

        # NAT Gateways (one per AZ for high availability)
        self.nat_gateways = []
        for i, subnet in enumerate(self.public_subnets):
            # Elastic IP for NAT Gateway
            eip = Eip(
                self,
                f"nat-eip-{i}",
                domain="vpc",
                tags={
                    "Name": f"dr-nat-eip-{i}-{region}-{environment_suffix}"
                }
            )

            # NAT Gateway
            nat = NatGateway(
                self,
                f"nat-gateway-{i}",
                allocation_id=eip.id,
                subnet_id=subnet.id,
                tags={
                    "Name": f"dr-nat-{i}-{region}-{environment_suffix}"
                }
            )
            self.nat_gateways.append(nat)

        # Private route tables (one per AZ)
        for i, (subnet, nat) in enumerate(zip(self.private_subnets, self.nat_gateways)):
            private_rt = RouteTable(
                self,
                f"private-route-table-{i}",
                vpc_id=self.vpc.id,
                route=[RouteTableRoute(
                    cidr_block="0.0.0.0/0",
                    nat_gateway_id=nat.id
                )],
                tags={
                    "Name": f"dr-private-rt-{i}-{region}-{environment_suffix}"
                }
            )

            RouteTableAssociation(
                self,
                f"private-rt-assoc-{i}",
                subnet_id=subnet.id,
                route_table_id=private_rt.id
            )

        # Security Groups

        # Lambda security group
        self.lambda_security_group = SecurityGroup(
            self,
            "lambda-sg",
            name=f"dr-lambda-sg-{region}-{environment_suffix}",
            description="Security group for Lambda functions",
            vpc_id=self.vpc.id,
            ingress=[],
            egress=[SecurityGroupEgress(
                from_port=0,
                to_port=0,
                protocol="-1",
                cidr_blocks=["0.0.0.0/0"],
                description="Allow all outbound traffic"
            )],
            tags={
                "Name": f"dr-lambda-sg-{region}-{environment_suffix}"
            }
        )

        # Aurora security group
        self.aurora_security_group = SecurityGroup(
            self,
            "aurora-sg",
            name=f"dr-aurora-sg-{region}-{environment_suffix}",
            description="Security group for Aurora database",
            vpc_id=self.vpc.id,
            ingress=[SecurityGroupIngress(
                from_port=5432,
                to_port=5432,
                protocol="tcp",
                security_groups=[self.lambda_security_group.id],
                description="PostgreSQL from Lambda"
            )],
            egress=[SecurityGroupEgress(
                from_port=0,
                to_port=0,
                protocol="-1",
                cidr_blocks=["0.0.0.0/0"],
                description="Allow all outbound traffic"
            )],
            tags={
                "Name": f"dr-aurora-sg-{region}-{environment_suffix}"
            }
        )
